File: sdk-java/src/main/java/io/littlehorse/sdk/wfsdk/NodeOutput.java
package io.littlehorse.sdk.wfsdk;

/**
 * `NodeOutput` represents the output of a Node execution. It can be used to set a timeout on a
 * node, or as input for a Variable Mutation.
 */
public interface NodeOutput extends LHExpression {
    /**
     * Valid only for output of the JSON_OBJ or JSON_ARR types. Returns a new NodeOutput handle
     * which points to Json element referred to by the json path.
     *
     * <p>This method is most often used to create the `rhs` parameter for
     * `ThreadBuilder::mutate()`.
     *
     * <p>Can only be called once--you can't call node.jsonPath().jsonPath().
     *
     * @param path is the json path to evaluate.
     * @return a NodeOutput.
     */
    public NodeOutput jsonPath(String path);

    /**
     * Valid only for output of JSON_OBJ or Struct types. Returns a new NodeOutput handle
     * which points to the sub-element referred to by the field.
     *
     * @param field is the name of the field to access.
     * @return a NodeOutput.
     */
    public NodeOutput get(String field);

    /**
     * Returns a new LHExpression that represents this node output cast to the specified type.
     * This enables manual type conversions that are not automatic, such as:
     * - STR → INT/DOUBLE/BOOL (manual casting required)
     * - DOUBLE → INT (manual casting required)
     *
     * The cast operation is non-mutating: the original node output remains unchanged.
     *
     * @param targetType the type to cast this node output to
     * @return a new LHExpression representing the cast value
     */
    // cast methods are provided by LHExpression
}

File: sdk-dotnet/LittleHorse.Sdk/Workflow/Spec/NodeOutput.cs
namespace LittleHorse.Sdk.Workflow.Spec;

/// <summary>
/// <c>NodeOutput</c> represents the output of a Node execution. It can be used to set a timeout on a
/// node, or as input for a Variable Mutation.
/// </summary>
public class NodeOutput
{
    /// <value>
    /// The node name of the output.
    /// </value>
    public string NodeName { get; private set; }
    
    /// <value>
    /// The workflow thread that this output belongs to.
    /// </value>
    public WorkflowThread Parent { get; private set; }
    
    /// <value>
    /// The Json path to the output. This is valid only for JSON_OBJ or JSON_ARR types.
    /// </value>
    public string? JsonPath { get; private set; }

    
    /// <summary>
    /// Initializes a new instance of the <see cref="NodeOutput"/> class with the specified node name
    /// and parent workflow thread.
    /// </summary>
    /// <param name="nodeName">The node name.</param>
    /// <param name="parent">The workflow thread that this output belongs to.</param>
    public NodeOutput(string nodeName, WorkflowThread parent)
    {
        NodeName = nodeName;
        Parent = parent;
    }
    
    /// <summary>
    /// Valid only for output of the JSON_OBJ or JSON_ARR types. Returns a new NodeOutput handle
    /// which points to Json element referred to by the json path.
    ///
    /// <p>This method is most often used to create the <c>rhs</c> parameter for
    /// <c>ThreadBuilder::Mutate()</c>. </p>
    ///
    /// <p>Can only be called once--you can't call <c>node.jsonPath().jsonPath().</c></p>
    /// </summary>
    /// <param name="path">
    /// Path is the json path to evaluate.
    /// </param>
    /// <returns>A NodeOutput.</returns>
    public NodeOutput WithJsonPath(string path) 
    {
        if (JsonPath != null) 
        {
            throw new Exception("Cannot use jsonpath() twice on same node!");
        }
        var nodeOutput = new NodeOutput(NodeName, Parent)
        {
            JsonPath = path
        };

        return nodeOutput;
    }
    
    /// <summary>
    /// Adds a timeout to a Node. Valid on TaskRuns and ExternalEvents.
    /// </summary>
    /// <param name="timeoutSeconds">
    /// The timeout length.
    /// </param>
    /// <returns>The NodeOutput.</returns>
    public NodeOutput WithTimeout(int timeoutSeconds) 
    {
        Parent.AddTimeoutToExtEvt(this, timeoutSeconds);
        
        return this;
    }
}